1309 create new UI for bug reporting phase that allows deleting and un deleting of issues

yucongkoo opened the pull request:
@Eclipse-Dominator @nknguyenhc @Arif-Khalid 
This is quite a big PR so it may require additional time for a thorough review. I appreciate your time and effort in reviewing it!

### Summary:
Fixes #1309 

### Changes Made:
* Implemented the new UI with 2 tables for bug reporting phase (and the bug-trimming phase will be using the same UI)
* Implemented the logic to display opened issues in the first table and closed issue in the second table, in short is I made the `issues.service` load all open and closed issues in the `initializeData` function, then I inject filtering based on open/closed status to the 2 `IssuesTableComponent`
* Updated the shared `IssuesTableComponent` to be able to accommodate a new restore issue icon with functionality of restoring deleted issues

### Proposed Commit Message:
Implement new UI and logic for bug reporting phase

As we are introducing a new bug-trimming phase 
for students to trim their bugs after the bug-reporting 
phase, having a single table of issues posted is not 
sufficient.

Let's implement a new UI where users can view
2 separate tables (one for open and the other for
closed issues), while allowing users to restore 
closed issues from the closed issues table too.

yucongkoo commented on src/app/core/services/permission.service.ts:
```diff
@@ -158,6 +158,11 @@ export class PermissionService {
     return this.askForPermission(PermissionLevel.User, 'isIssueDeletable');
   }
 
+  // should only have restore permission if have delete permission
+  isIssueRestorable(): boolean {
+    return this.isIssueDeletable();
+  }
+
```
Instead of having a new permission config, I thought that this would make sense, since deleting and restoring are basically opposite actions (one undo-ing the other), their permission should be the same too. I would love to listen inputs on this decision.

yucongkoo commented on src/app/shared/issue-tables/issue-tables.component.ts:
```diff
@@ -45,6 +46,7 @@ export class IssueTablesComponent implements OnInit, AfterViewInit {
 
   issues: IssuesDataTable;
   issuesPendingDeletion: { [id: number]: boolean };
+  issuesPendingRestore: { [id: number]: boolean };
```
This is a placeholder so that I don't miss this (can be removed after discussion).

Based on the current implementation of deleting issues, it seems like there is a mechanism to hold the issues pending deletion to prevent users from spamming delete on the same issue, I was thinking we should probably implement a similar mechanism for restoring issues too?

If we want to I will create a separate issue for this.

yucongkoo reviewed (commented):
I have included some places where I am not super sure about my change (there might be better ways to implement this that I couldn't think of).

yucongkoo commented:
Also, I have also observed that there is high code duplication in the `issues-tables.component.ts` for the functions `deleteIssue` and `undeleteIssue`, I also don't quite like the fact that there is a lot of logic within the `issues-tables.component.html` similar to the following
<img width="1112" alt="Screenshot 2025-02-05 at 10 45 42 PM" src="https://github.com/user-attachments/assets/54928985-95ad-4cc5-91d0-3db5ed912772" />

But I don't want to do too much refactoring as this would increase the difficulty of reviewing the PR. Hence I plan to create a separate Issue to work on refactoring these issues.

Eclipse-Dominator commented on src/app/core/services/issue.service.ts:
```diff
@@ -321,18 +320,23 @@ export class IssueService {
     const issuesAPICallsByFilter: Array<Observable<Array<GithubIssue>>> = [];
 
+    let filter: RestGithubIssueFilter = new RestGithubIssueFilter({});
+    if (this.phaseService.requireLoadClosedIssues()) {
+      filter.state = 'all';
+    }
```
I think this is fine actually, however, the way this func is structured together with the inner functions makes it less efficient and harder to read. (So not an issue here but can be optimized for better clarity and efficiency in the future. See below) (Refer to #1313)

Eclipse-Dominator commented on src/app/shared/issue-tables/issue-tables.component.ts:
```diff
@@ -180,41 +176,54 @@ export class IssueTablesComponent implements OnInit, AfterViewInit {
+    }
+
     let snackBarRef = null;
     snackBarRef = this.snackBar.openFromComponent(UndoActionComponent, {
       data: { message: `Deleted issue ${id}` },
       duration: this.snackBarAutoCloseTime
     });
     snackBarRef.onAction().subscribe(() => {
-      this.undeleteIssue(id, event);
+      this.undeleteIssue(id, event, false);
```
I know this code is originally here.
But I don't think displaying it immediately on click.
We need to consider failing cases should GitHub fail to delete this. 
Thus, it might not be wise to immediately display this when delete button is clicked but only when the callback from `undelete` and `delete` returns the new issues (which indicate a successful modification in the Github server.

yucongkoo replied:
Addressed

Eclipse-Dominator commented on src/app/shared/issue-tables/issue-tables.component.ts:
```diff
@@ -180,41 +176,54 @@ export class IssueTablesComponent implements OnInit, AfterViewInit {
-        this.logger.info(`IssueTablesComponent: Deleting issue ${id}`);
-        this.deleteIssue(id, event);
-      }
+    let snackBarRef = null;
+    snackBarRef = this.snackBar.openFromComponent(UndoActionComponent, {
+      data: { message: `Restored issue ${id}` },
+      duration: this.snackBarAutoCloseTime
+    });
+    snackBarRef.onAction().subscribe(() => {
+      this.deleteIssue(id, event, false);
```
same as above

yucongkoo replied:
Addressed

Eclipse-Dominator reviewed (changes requested):
Good work.

Some points to consider: maybe dont use the word 'delete'. I feel like its a bit too strong here (just my own opinion). maybe hide might be better? 

We can just tell them the bug restored are just hidden and can be restored during bug trimming phases. 
I dont think having the 2nd table with the concept of delete and restore adds much value to the users during the bug reporting phase UNLESS hidden bugs will not be transferred to the bug trimming phase.

So, this is mainly just a visual thing for the bug reporting phase.

Eclipse-Dominator commented on src/app/core/services/issue.service.ts:
```diff
@@ -265,7 +264,7 @@ export class IssueService {
       map((response: GithubIssue) => {
         const reopenedIssue = this.createIssueModel(response);
-        this.updateLocalStore(reopenedIssue);
+        this.deleteFromLocalStore(reopenedIssue);
```
made an issue regarding in #1314

Eclipse-Dominator replied:
In case you are not sure what to do here, I recommend making both `deleteIssue` and `undeleteIssue` to call `updateLocalStore`.

Fix the above bug as a follow up in a seperate PR.

yucongkoo replied:
Addressed.

yucongkoo commented on src/app/core/services/issue.service.ts:
```diff
@@ -87,15 +87,14 @@ export class IssueService {
-          );
-        }
-      )
+          map((response) => {
+            const issue = this.createIssueModel(response);
+            this.updateLocalStore(issue);
+            return issue;
+          }),
+          catchError((err) => this.getIssue(issueId))
+        );
+      })
```
This was not a IDE formatting actually, rather it was one of the pre-commit hooks being ran by husky, and the only way to bypass this (at least from what I know) is via the command `git commit --no-verify`. 

The following is what happens when I try to commit the file where I reverted the changes, and the formatting is being fixed to the new formatting once again (at commit time the file was already saved as the old formatting)
<img width="982" alt="Screenshot 2025-02-08 at 4 37 18 PM" src="https://github.com/user-attachments/assets/3e178eb1-ab69-429b-8917-7e9e5a6ab959" />

Personally I think for this case it would be better to use the fixed formatting instead, what do you think?

yucongkoo commented:
@Eclipse-Dominator  I have addressed the issues mentioned above!

> We can just tell them the bug deleted are just hidden and can be restored during bug trimming phases. I dont think having the 2nd table with the concept of delete and restore adds much value to the users during the bug reporting phase UNLESS hidden bugs will not be transferred to the bug trimming phase.

Yeah this design of 2 tables was the design we thought of for the bug trimming phase, but prof Damith suggested that we could make these 2 phases to use the same design since it just provides more flexibility, we might want to have a more in-depth discussion on how the UI/UX for these 2 phases.

Currently the bug-trimming phase is still being set up, so I think implementing this now would be beneficial to just port the code over for bug-trimming phase anyways(and we can discover potential issues earlier too).

Eclipse-Dominator commented on src/app/shared/issue-tables/issue-tables.component.ts:
```diff
@@ -158,63 +154,69 @@ export class IssueTablesComponent implements OnInit, AfterViewInit {
-          this.issuesPendingDeletion = theRest;
-        })
-      )
+      .pipe(finalize(() => this.handleIssueDeletionFinalization(id, event, actionUndoable)))
```
Do correct me if I am wrong, I believe finalize would still be called should there be an error.  
So should Github fail to close/open an issue, the undo option will still exist.

Thus, let's not introduce additional complexities.

This can be combined into 
```typescript
      .subscribe(
        (removedIssue) => this.handleIssueDeletionFinalization(id, event, actionUndoable),
        (error) => this.errorHandlingService.handleError(error)
      );
```

yucongkoo replied:
Ahhh this was my bad, I wasn't aware that `finalize` worked like that.

After understanding more on how `finalize` and `subscribe` works, I believe we would still need to use `finalize` here for the logic that removes the issues from `issuesPendingDeletion` list (if not when the deletion fails, the UI will continuous display the spinner icon), I have refactored the code block as follows:
```
this.issueService
  .deleteIssue(id)
  .pipe(finalize(() => {
    const { [id]: issueRemoved, ...theRest } = this.issuesPendingDeletion;
    this.issuesPendingDeletion = theRest;
  }))
  .subscribe(
    (removedIssue) => this.handleIssueDeletionSuccess(id, event, actionUndoable),
    (error) => this.errorHandlingService.handleError(error) 
  );
```

Eclipse-Dominator commented on src/app/shared/issue-tables/issue-tables.component.ts:
```diff
@@ -158,63 +154,69 @@ export class IssueTablesComponent implements OnInit, AfterViewInit {
+    this.issuesPendingRestore = { ...this.issuesPendingRestore, [id]: true };
+    this.issueService
+      .undeleteIssue(id)
+      .pipe(finalize(() => this.handleIssueRestorationFinalization(id, event, actionUndoable)))
+      .subscribe(
+        (restoredIssue) => {},
+        (error) => {
+          this.errorHandlingService.handleError(error);
+        }
```
Same as above

This can be combined into:

```suggestion
      .subscribe(
        (restoredIssue) => this.handleIssueRestorationFinalization(id, event, actionUndoable),
        (error) => this.errorHandlingService.handleError(error)
      );
    event.stopPropagation();
```

yucongkoo replied:
Same as above

Eclipse-Dominator commented on src/app/shared/issue-tables/issue-tables.component.ts:
```diff
@@ -158,63 +154,69 @@ export class IssueTablesComponent implements OnInit, AfterViewInit {
   }
 
-  deleteIssue(id: number, event: Event) {
+  private handleIssueDeletionFinalization(id: number, event: Event, actionUndoable: boolean) {
+    const { [id]: issueRemoved, ...theRest } = this.issuesPendingDeletion;
+    this.issuesPendingDeletion = theRest;
+
+    if (!actionUndoable) {
+      return;
+    }
+    let snackBarRef = null;
+    snackBarRef = this.snackBar.openFromComponent(UndoActionComponent, {
+      data: { message: `Deleted issue ${id}` },
+      duration: this.snackBarAutoCloseTime
+    });
+    snackBarRef.onAction().subscribe(() => {
+      this.undeleteIssue(id, event, false);
+    });
+  }
```
I think this is fine for now.

Do consider if issuesPendingDeletion and Restore are a good design choice for this case.

yucongkoo replied:
Yes I was already aware of this design choice and already created an [issue](https://github.com/CATcher-org/CATcher/issues/1315) that I plan to work on after this.

Eclipse-Dominator reviewed (changes requested):
Some minor refactoring and its good to go!

yucongkoo commented on src/app/core/services/issue.service.ts:
```diff
@@ -87,15 +87,14 @@ export class IssueService {
-        }
-      )
+          map((response) => {
+            const issue = this.createIssueModel(response);
+            this.updateLocalStore(issue);
+            return issue;
```
Yup I agree, I verified that the return value is not used anywhere.

yucongkoo commented:
@Eclipse-Dominator 

Thanks for pointing out the possible refactoring. I have improved the mentioned areas!

Eclipse-Dominator commented:
@yucongkoo I see you used createAndSaveIssueModel to refactor some code!
 I think the same function can also replace the callback in `deleteIssue` and `undeleteIssue`.

Eclipse-Dominator reviewed (approved):
LGTM! Good work!
